yktrace/src/tir.rs

```diff
@@ -285,74 +258,143 @@ impl TirTrace {
+                assert!(s.contains("stop_tracing"))
+            }
+            e => panic!("Expected call to `stop_tracing` here, instead got {:?}.", e)
+        }
+        match ops.pop() {
+            Some(TirOp::Statement(Statement::Assign(_, _))) => {
+            }
+            e => panic!("Expected `Assign` here, instead got {:?}.", e)
+        }
+        match ops.pop() {
+            Some(TirOp::Statement(Statement::Assign(_, Rvalue::Use(Operand::Constant(Constant::Bool(false)))))) => {
+            }
+            e => panic!("Expected `Assign(_, false)` here, instead got {:?}.", e)
+        }
+        for _ in (0..3) {
+            match ops.pop() {
+                Some(TirOp::Statement(Statement::StorageLive(_))) => {
+                }
+                e => panic!("Expected `StorageLive` here, instead got {:?}.", e)
+            }
+        }
+        match ops.remove(0) {
+            TirOp::Statement(Statement::StorageDead(_)) => {
+            }
+            e => panic!("Expected `StorageDead` here, instead got {:?}.", e)
+        }
+
         Ok(Self { ops })
     }
 
-    fn rename_rvalue(rename_map: &HashMap<Local, Vec<Local>>, rvalue: &Rvalue) -> Rvalue {
+    /// Return the TIR operation at index `idx` in the trace.
+    /// The index must not be out of bounds.
+    pub fn op(&self, idx: usize) -> &TirOp {
+        debug_assert!(idx <= self.ops.len() - 1, "bogus trace index");
+        unsafe { &self.ops.get_unchecked(idx) }
+    }
+
+    /// Return the length of the trace measure in operations.
+    pub fn len(&self) -> usize {
+        self.ops.len()
+    }
+}
+
+struct SSA {
+    /// Largest original variable name at the time of entering a call.
+    stack: Vec<u32>,
+    /// Current offset used to rename variables.
+    offset: u32,
+}
+
+impl SSA {
+    fn new() -> Self {
+        SSA {
+            stack: vec![0],
+            offset: 0,
+        }
+    }
+
+    fn offset(&self) -> u32 {
+        self.offset
+    }
+
+    fn enter(&mut self, num_locals: usize) {
+        // When entering an inlined function call we push the current offset onto the stack, so we
+        // can return to this offset after leaving the function. Afterwards we increase the offset
+        // by the number of variables in the outer context.
+        self.offset += num_locals as u32;
+        self.stack.push(self.offset);
+    }
+
+    fn leave(&mut self) {
+        // When we leave an inlined function call, we pop the previous offset from the stack,
+        // reverting the offset to what it was before the function was entered.
+        self.stack.pop();
+        if let Some(v) = self.stack.last() {
+            self.offset = *v;
+        } else {
+            panic!("Unbalanced enter/leave statements!")
+        }
+    }
+
+    fn rename_args(&mut self, args: &Vec<Operand>) -> Vec<Operand> {
+        let mut newargs = Vec::new();
+        for (i, op) in args.iter().enumerate() {
```

Comment:
We might be able to make it slightly more concise with `.map()`? (but I wouldn't worry too much).

Reply:
This used to do more stuff, but has been simplified since then, so I agree this can be a map now.

Reply:
Fixed in 0f6e990